pkg/ruler/rulestore/config.go

```diff
@@ -0,0 +1,58 @@
+package rulestore
+
+import (
+	"context"
+	"flag"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/cortexproject/cortex/pkg/configs/client"
+	"github.com/cortexproject/cortex/pkg/ruler/rules"
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+)
+
+const (
+	ConfigDB = "configdb"
+
+	name   = "rule-store"
+	prefix = "ruler-storage."
+)
+
+// Config configures a rule store.
+type Config struct {
+	bucket.Config `yaml:",inline"`
+	ConfigDB      client.Config `yaml:"configdb"`
+}
+
+// RegisterFlags registers the backend storage config.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.ExtraBackends = []string{ConfigDB}
+	cfg.ConfigDB.RegisterFlagsWithPrefix(prefix, f)
+	cfg.RegisterFlagsWithPrefix(prefix, f)
+}
+
+// NewRuleStore creates a new bucket client based on the configured backend
+func NewRuleStore(ctx context.Context, cfg Config, logger log.Logger, reg prometheus.Registerer) (rules.RuleStore, error) {
+	if cfg.Backend == ConfigDB {
```

Comment:
Why do we need another way of configuring obsolete ConfigDB rule store? It's already possible to use it without having this option.

Reply:
Based on [this discussion](https://github.com/cortexproject/cortex/discussions/3771#discussioncomment-327738) we would eventually deprecate the config options currently configured using the `-ruler.storage.*` flag prefix in favor of the the newly added config field. However, the config service is still in use and needs to be carried over. I'm in favor of deprecating the config service but I think it would be best to scope that separately.

Reply:
I didn't mean to deprecate configdb (although perhaps we should), just that we don't need a separate way of configuring it.

The way I was thinking about the transition was to add config options to existing `RuleStoreConfig`, and adding new types like "bucket-s3". But I see that it could create more confusion.

Ideally we will do this without breaking changes for ConfigDB users.

Reply:
> The way I was thinking about the transition was to add config options to existing RuleStoreConfig, and adding new types like "bucket-s3". But I see that it could create more confusion.

I also felt it would be more confusing to try and fit this change into the existing config struct instead of creating a new top level struct. As for the configdb, the upgrade path would be identical to the bucket upgrade path, except only the prefix of the flags would change.

However, another option would be to:
1. Not add the configdb as an option for the new RuleStore config struct
2. When deprecating the existing objectclient flags/fields, maintain the config options for the `configdb`. However, give a clear warning that this client is not recommended for new deployments.
3. Figure out the future of the configdb service separately